fix: add filters info in footnote

jjxie opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/qlik-oss/nebula.js/blob/master/.github/CONTRIBUTING.md#git
-->

## Motivation

Added support of showing filters information in footnote. 
`generateSetExpression.js` is same check for classic chart

**None card theme**
![Screenshot 2025-04-29 at 15 58 30](https://github.com/user-attachments/assets/2430ef9d-cffa-459b-9c97-19eb3f797337)

**Card theme**
![Screenshot 2025-04-29 at 15 58 51](https://github.com/user-attachments/assets/d5a580a8-2787-439b-b24d-2a248d9ab2c7)



## Requirements checklist

<!-- Make sure you got these covered -->

- [ ] Api specification
  - [ ] Ran `yarn spec`
    - [ ] No changes **_OR_** API changes has been formally approved
- [ ] Unit/Component test coverage
- [ ] Correct PR title for the changes (fix, chore, feat)

When build and tests have passed:

- [ ] Add code reviewers, for example @qlik-oss/nebula-core

T-Wizard commented:
wrong footer size when theme is a card theme and `VNA-13_CELLPADDING_FROM_THEME` is on

Caele commented on apis/nucleus/src/utils/generateSetExpression.js:
```diff
@@ -0,0 +1,234 @@
+const FilterType = {
```
This file should have tests, I see that there are some for the original one, so bring them over.

jjxie commented:
> wrong footer size when theme is a card theme and `VNA-13_CELLPADDING_FROM_THEME` is on

I added padding and border top style for filter footnote with card theme. 

Removed padding top when new sheet flag is enabled (previously 8px). Padding top for non card theme, like sense classic is 6px, for card theme is 0px. I am not sure if it is always the case, or can change padding top in themes. I noticed font size and style is not applied for some themes, both for normal footnote text or filter information. I am not sure if it is related to your changes of applying theme. I will check, but will not make change in this PR.

jjxie commented on apis/nucleus/src/components/listbox/utils/frequencyMaxUtil.js:
```diff
@@ -1,12 +1,4 @@
-const escapeField = (field) => {
-  if (!field) {
```
This line the new file check `if (!field || field === ']')`. I guess it is fine to use new file logic. @T-Wizard

T-Wizard replied:
OK.
if I remember correctly `field === ']'` was from engine making it impossible to use a field with the name `]`
but now it looks like it works both with and without escape

T-Wizard commented on apis/nucleus/src/utils/cell-padding.js:
```diff
@@ -41,7 +41,8 @@ const getPadding = ({ layout, isError, senseTheme, titleStyles }) => {
     const showTitle = layout?.showTitles && !!layout?.title;
     const showSubtitle = layout?.showTitles && !!layout?.subtitle;
-    const showFootnote = layout?.showTitles && !!layout?.footnote;
+    const hasFilters = layout?.filters?.length > 0 && layout?.qHyperCube?.qMeasureInfo?.length > 0;
+    const showFootnote = layout?.showTitles && (!!layout?.footnote || hasFilters);
```
also need to do `const filtersFootnoteString = generateFiltersString(layout?.filters ?? [], translator);` and check if filtersFootnoteString is empty string or not

T-Wizard commented on apis/nucleus/src/components/Footer.jsx:
```diff
@@ -30,15 +29,21 @@ const CellFooter = {
+        item
+        className={classes.itemStyle}
+        data-testid={CellFooter.className}
+        style={{ paddingTop: (theme) => (newSheetEnabled ? 0 : theme.spacing(1)) }}
```
from my testing.
For classic theme (not cards) it should always have spacing.
and for cards theme 0 is to small, maybe 1px?

T-Wizard commented on apis/nucleus/src/components/FiltersFooter.jsx:
```diff
@@ -0,0 +1,63 @@
+        <Grid item display="flex">
+          {filtersFootnoteLabels.map((filter) => (
+            <Grid container wrap="nowrap" key={`${filter.field}-${filter.label}`}>
+              <BoldText styles={styles}>{`${filter.field}:`}</BoldText>
```
when I compare with sense-client this should be bold and italic

Caele commented on apis/nucleus/src/components/FiltersFooter.jsx:
```diff
@@ -0,0 +1,55 @@
+          {filtersFootnoteLabels.map((filter) => (
+            <Grid container wrap="nowrap" key={`${filter.field}-${filter.label}`}>
+              <ItalicText styles={{ ...styles, fontWeight: 'bold' }}>{`${filter.field}:`}</ItalicText>
+              <ItalicText styles={styles}> &nbsp;</ItalicText>
+              <ItalicText styles={styles}> {filter.label}</ItalicText>
+              <ItalicText styles={styles}> &nbsp;</ItalicText>
```
These 3 all have the same style, can't they be a single on instead?

jjxie replied:
Done

Caele commented on apis/nucleus/src/components/Footer.jsx:
```diff
@@ -28,15 +29,37 @@ const CellFooter = {
+function Footer({ layout, titleStyles = {}, translator, flags, isCardTheme }) {
+  const footerStyle = titleStyles.footer;
+  const hasFilters = layout?.filters?.length > 0 && layout?.qHyperCube?.qMeasureInfo?.length > 0;
+  const filtersFootnoteString = generateFiltersString(layout?.filters ?? [], translator);
```
You have these two functions: generaFiltersString and generateFiltersLabels, but one is called out here in Footer and one inside FiltersFooter. I think it would be cleaner if all that is done inside the FiltersFooter.
I see that you need it to check if the FiltersFooter should exist or not, but you can do that check inside the component instead, which would clean up the Footer quite a bit.

Caele replied:
its ok

Caele commented on apis/nucleus/src/utils/generateSetExpression.js:
Most of this file isn't used right? So we should not keep those things here. Better clean it out together with those tests.

jjxie replied:
Most of this file is used, but `generateSetExpression` function is not used. I removed this function and moved some tests were testing `generateSetExpression` before to test label strings.
